### PR TITLE
HIP: Initialize device-related variables only by the singleton

### PR DESCRIPTION
**Pull request by masterleinad**
Analogous to #5442 for HIP.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP.cpp [hunk 1]
```diff
@@ -65,10 +66,73 @@ int HIP::impl_is_initialized() {
 }
 
 void HIP::impl_initialize(InitializationSettings const& settings) {
+  const int hip_device_id = ::Kokkos::Impl::get_gpu(settings);
+
+  const auto& dev_info = Impl::HIPInternalDevices::singleton();
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP.cpp:71]: If we are refactoring we should just get rid of that `Impl::HIPInternalDevices` altogether

File: core/src/HIP/Kokkos_HIP.cpp [hunk 2]
```diff
@@ -65,10 +66,71 @@ int HIP::impl_is_initialized() {
 }
 
 void HIP::impl_initialize(InitializationSettings const& settings) {
+  const int hip_device_id = ::Kokkos::Impl::get_gpu(settings);
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP.cpp:69]:
```suggestion
  const int hip_device_id = Impl::get_gpu(settings);
```

File: core/src/HIP/Kokkos_HIP.cpp [hunk 3]
```diff
@@ -65,10 +66,71 @@ int HIP::impl_is_initialized() {
...
+
+  // Need at least a GPU device
+  int hipDevCount;
+  KOKKOS_IMPL_HIP_SAFE_CALL(hipGetDeviceCount(&hipDevCount));
+  const bool ok_id = 0 <= hip_device_id && hip_device_id < hipDevCount;
+
+  if (ok_id) {
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP.cpp:76]: This is unnecessary.  The `device_id` validation is done by Kokkos Core

File: core/src/HIP/Kokkos_HIP.cpp [hunk 4]
```diff
@@ -65,10 +66,71 @@ int HIP::impl_is_initialized() {
...
+  } else {
+    std::ostringstream msg;
+    msg << "Kokkos::HIP::initialize(" << hip_device_id
+        << ") FAILED : Device identifier out of range "
+        << "[0.." << hipDevCount - 1 << "]";
+    Kokkos::Impl::throw_runtime_exception(msg.str());
+  }
+
+  // Init the array for used for arbitrarily sized atomics
+  ::Kokkos::Impl::initialize_host_hip_lock_arrays();
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP.cpp:120]:
```suggestion
  Impl::initialize_host_hip_lock_arrays();
```

File: core/src/HIP/Kokkos_HIP_Instance.cpp [hunk 5]
```diff
@@ -87,37 +87,19 @@ Kokkos::View<uint32_t *, HIPSpace> hip_global_unique_token_locks(
...
-}
-
-const HIPInternalDevices &HIPInternalDevices::singleton() {
-  static HIPInternalDevices self;
-  return self;
-}
-}  // namespace
 
+int Impl::HIPInternal::m_hipDev                                           = -1;
+int Impl::HIPInternal::m_hipArch                                          = -1;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Instance.cpp:92]: Declare them `inline` and define them inside the `HIPInternal` class definition instead?

File: core/src/HIP/Kokkos_HIP_Instance.hpp [hunk 6]
```diff
@@ -121,7 +121,7 @@ class HIPInternal {
   size_type *m_scratchSpace = nullptr;
   size_type *m_scratchFlags = nullptr;
 
-  hipDeviceProp_t m_deviceProp;
+  static hipDeviceProp_t m_deviceProp;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Instance.hpp:124]: Readability, move it next to the other static members

**Review by dalg24 — CHANGES_REQUESTED** [review]
Note that it likely conflicts with #5448

File: core/src/HIP/Kokkos_HIP.cpp [hunk 7]
```diff
@@ -65,10 +66,61 @@ int HIP::impl_is_initialized() {
 }
 
 void HIP::impl_initialize(InitializationSettings const& settings) {
+  const int hip_device_id = Impl::get_gpu(settings);
+
+  // Need at least a GPU device
+  int hipDevCount;
+  KOKKOS_IMPL_HIP_SAFE_CALL(hipGetDeviceCount(&hipDevCount));
```
Comment (ldh4) [inline core/src/HIP/Kokkos_HIP.cpp:73]: Seeing that `HIP::detect_device_count()` now does the same thing as this line,

```suggestion
  int hipDevCount = HIP::detect_device_count();
```

**Comment by masterleinad** [comment]
THE HIP CI builds pass. A couple of CUDA CI steps fail because the tester was running out of disk space.

**Comment by masterleinad** [comment]
Updated after #5448 was merged.

**Review by dalg24 — APPROVED** [review]
Fine by me but requiring approval of Bruno

**Review by Rombur — APPROVED** [review]
Looks fine but you need to fix the conflict.

File: core/src/HIP/Kokkos_HIP_Instance.cpp (hunk 5 above)
Comment (dalg24) [reply core/src/HIP/Kokkos_HIP_Instance.cpp:92]: Ping

**Comment by masterleinad** [comment]
> Looks fine but you need to fix the conflict.

Merged with `develop` again...

**Comment by Rombur** [comment]
This can be merged. The CI failures are clearly unrelated.

**Pull request merged**